subsys/event_manager/event_manager.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <spinlock.h>
+#include <sys/slist.h>
+#include <event_manager/event_manager.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(event_manager, CONFIG_EVENT_MANAGER_LOG_LEVEL);
+
+
+/* Event manager uses orphan sections. Below tag will allow linker to
+ * find a section usable to put the orphan sections next to.
+ */
+const struct {} linker_tag __attribute__((__section__("event_manager"))) __used;
+
+
+static void event_processor_fn(struct k_work *work);
+
+struct event_manager_event_display_bm _event_manager_event_display_bm;
+
+static K_WORK_DEFINE(event_processor, event_processor_fn);
+static sys_slist_t eventq = SYS_SLIST_STATIC_INIT(&eventq);
+static struct k_spinlock lock;
+
+
+static bool log_is_event_displayed(const struct event_type *et)
+{
+	return atomic_test_bit(_event_manager_event_display_bm.flags, et - __start_event_types);
+}
+
+static void log_event(const struct event_header *eh)
+{
+	const struct event_type *et = eh->type_id;
+
+	if (!IS_ENABLED(CONFIG_EVENT_MANAGER_SHOW_EVENTS) ||
+	    !log_is_event_displayed(et)) {
+		return;
+	}
+
+	if (et->log_event) {
+		char log_buf[CONFIG_EVENT_MANAGER_EVENT_LOG_BUF_LEN];
```

Comment:
This way of logging is not very optimal (sprintf + logging strings are much slower than logging values). Wonder if it could be reworked so that event logging function is logging directly.
I will come up with a proposal just need to check it.

Reply:
here is my proposal:
https://github.com/nordic-krch/zephyr/commits/event_manager_logging
you can find two commits there, one with logging proposal. It enables logging directly using logging API and not doing sprintf. In second commit i tried to compare CPU load impact. Calculated load for things other than idle and logging thread to get load of application. I got `0.45%` with new logging and `0.66%` without. So seems like app was spending 33% of time sprintf-logging.

Reply:
Thanks for your input. I thought that it should be backward compatible with the previous version.  Do you thing Kconfig option would be ok in this case?

Reply:
it could be but since current version does not have any benefits over this proposal it should be eventually deprecated and removed. Prototype of logging function shall be changed when newer version is used.

Reply:
Please don't reuse the old name `log_event`. We should create a new member of the structure.

This is simple change (our shame we have not done macros hiding logging in the first place when we migrated the event manager out of nrf_desktop), bigger ones are introduced by zephyr every quarter.
Nevertheless adding backward compatibility should be very simple. You can add a config option and handle log either through old `log_event` or new `log_fn` structure member. Config option could make only one of them available (preferred option since we would like to drop the old ways).

Reply:
Have another thought about it. We cannot do it in any other way then proposed by @nordic-krch . Adding a new member would required to change the macro defining an event (so we can assign right member with a given logging callback). Doing that conditionally based on kconfig option is not really safe as at some point there will be a conflict between function given and expected.
@nordic-krch proposal using the old api gives an option for explicit check on the compatibility by passing the null pointer as buffer. So let's do it as was proposed.

We could still keep the simple backward compatibility at event_manager.c - if enabled buffer is created and passed to the function as it used to be. New functions should handle it fine. @zycz , let's add this minimum backward compatibility option and inform in change log that old ways will be dropped.